ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/MiniProtocol/ObjectDiffusion/Inbound/V2/Types.hs

```diff
@@ -0,0 +1,325 @@
+  -- , dpsObjectsRequestedButNotReceivedIds :: !(Set objectId)
+  -- ^ A subset of `dpsOutstandingFifo` which were unknown to the peer
+  -- (i.e. requested but not received). We need to track these `objectId`s
+  -- since they need to be acknowledged.
+  , dpsObjectsOwtPool :: !(Map objectId object)
+  -- ^ A set of objects on their way to the objectpool.
+  -- Tracked here so that we can cleanup `dgsObjectsOwtPoolMultiplicities` if the
+  -- peer dies.
+  --
+  -- Life cycle of entries:
+  -- * added by `acknowledgeObjectIds` (where decide which objects can be
+  --   submitted to the objectpool)
+  -- * removed by `withObjectPoolSem`
+  }
+  deriving stock (Show, Eq, Generic)
+  deriving anyclass (NFData, NoThunks)
+
+-- | Shared state of all `ObjectDiffusion` clients.
+data DecisionGlobalState peerAddr objectId object = DecisionGlobalState
+  { dgsPeerStates :: !(Map peerAddr (DecisionPeerState objectId object))
+  -- ^ Map of peer states.
+  --
+  -- /Invariant:/ for peerAddr's which are registered using `withPeer`,
+  -- there's always an entry in this map even if the set of `objectId`s is
+  -- empty.
+  }
+  deriving stock (Show, Eq, Generic)
+  deriving anyclass (NFData, NoThunks)
+
+-- | Merge dpsObjectsAvailableIds from all peers of the global state.
+dgsObjectsAvailableMultiplicities ::
+  Ord objectId => DecisionGlobalState peerAddr objectId object -> Map objectId ObjectMultiplicity
+dgsObjectsAvailableMultiplicities DecisionGlobalState{dgsPeerStates} =
+  Map.unionsWith
+    (+)
+    (Map.fromSet (const 1) . dpsObjectsAvailableIds <$> Map.elems dgsPeerStates)
+
+dgsObjectsInflightMultiplicities ::
+  Ord objectId => DecisionGlobalState peerAddr objectId object -> Map objectId ObjectMultiplicity
+dgsObjectsInflightMultiplicities DecisionGlobalState{dgsPeerStates} =
+  Map.unionsWith
+    (+)
+    (Map.fromSet (const 1) . dpsObjectsInflightIds <$> Map.elems dgsPeerStates)
+
+dgsObjectsOwtPoolMultiplicities ::
+  Ord objectId => DecisionGlobalState peerAddr objectId object -> Map objectId ObjectMultiplicity
+dgsObjectsOwtPoolMultiplicities DecisionGlobalState{dgsPeerStates} =
+  Map.unionsWith
+    (+)
+    (Map.fromSet (const 1) . Map.keysSet . dpsObjectsOwtPool <$> Map.elems dgsPeerStates)
+
+type DecisionGlobalStateVar m peerAddr objectId object =
+  StrictTVar m (DecisionGlobalState peerAddr objectId object)
+
+newDecisionGlobalStateVar ::
+  MonadSTM m =>
+  m (DecisionGlobalStateVar m peerAddr objectId object)
+newDecisionGlobalStateVar =
+  newTVarIO
+    DecisionGlobalState
+      { dgsPeerStates = Map.empty
+      }
+
+--
+-- Decisions
+--
+
+data DecisionContext peerAddr objectId object = DecisionContext
+  { dcRng :: StdGen
+  , dcHasObject :: (objectId -> Bool)
+  , dcDecisionPolicy :: DecisionPolicy
+  , dcGlobalState :: DecisionGlobalState peerAddr objectId object
+  , dcPrevDecisions :: Map peerAddr (PeerDecision objectId object)
+  }
+  deriving stock Generic
+  deriving anyclass NFData
```

Comment:
Could you maybe add some documentation here?

Concretely, I would like to understand if we need both `DecisionGlobalState` and `DecisionContext`, as their names somewhat imply certain overlapping in their responsibilities.